### PR TITLE
Runtime logs

### PR DESCRIPTION
**Pull request by crufter**
All very experimental and comes with the following caveats:

- No support for structured logs yet. No metadata, no timestamp, just actual log.
- Local logging does not support past logs, only stream as they come
- K8s logs might be out of order at the start of the stream.

### CODE REVIEW

File: runtime/options.go [hunk 1]
```diff
@@ -149,3 +149,38 @@ func ReadType(t string) ReadOption {
...
+	ExistingCount int64
+	// Stream new lines?
+	Stream bool
+	// If streaming is on, how many lines to stream?
+	// 0 means keep streaming until the LogStreamer is closed.
+	StreamCount int64
+}
+
+// LogsExistingCount confiures how many existing lines to show
+func LogsExistingCount(count int64) LogsOption {
```
Comment (asim) [inline runtime/options.go]: LogCount

File: runtime/options.go [hunk 2]
```diff
@@ -149,3 +149,38 @@ func ReadType(t string) ReadOption {
...
+
+// LogsStream configures whether to stream new lines
+func LogsStream(stream bool) LogsOption {
+	return func(l *LogsOptions) {
+		l.Stream = stream
+	}
+}
+
+// LogsStream configures how many lines of lgos to stream
+func LogsStreamCount(streamCount int64) LogsOption {
```
Comment (asim) [inline runtime/options.go]: I would not specify how many messages to stream. Just that we support streaming.

File: runtime/runtime.go [hunk 3]
```diff
@@ -35,6 +35,19 @@ type Runtime interface {
...
+}
+
+// Stream returns a log stream
+type LogStream interface {
+	Chan() chan LogRecord
+	Stop() error
+}
+
+type LogRecord struct {
+	Log  string
```
Comment (asim) [inline runtime/runtime.go]: Message string

File: runtime/runtime.go [hunk 4]
```diff
@@ -35,6 +35,19 @@ type Runtime interface {
...
+	Meta map[string]string
```
Comment (asim) [inline runtime/runtime.go]: Metadata

File: runtime/kubernetes/kubernetes_logs.go [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+// NewLog returns a configured Kubernetes logger
+func newLog(serviceName string, opts ...runtime.LogsOption) *klog {
+	klog := &klog{
+		serviceName: serviceName,
+	}
+	for _, o := range opts {
+		o(&klog.options)
+	}
+
+	if len(os.Getenv("KUBERNETES_SERVICE_HOST")) > 0 {
```
Comment (asim) [inline runtime/kubernetes/kubernetes_logs.go]: Pass the existing client thats in the runtime to newLog

File: runtime/kubernetes/kubernetes_logs.go [hunk 6]
```diff
@@ -0,0 +1,170 @@
+// Package kubernetes taken from https://github.com/micro/go-micro/blob/master/debug/log/kubernetes/kubernetes.go
+// There are some modifications compared to the other files as
+// this package doesn't provide write functionality.
+// With the write functinality gone, structured logs also go away.
```
Comment (jakexks) [inline runtime/kubernetes/kubernetes_logs.go]:
```suggestion
// With the write functionality gone, structured logs also go away.
```

**Review by jakexks — COMMENTED** [review]
Looks OK to me!

**Review by jakexks — APPROVED** [review]

**Pull request merged**